iwyu_output.cc

```diff
@@ -578,14 +578,19 @@ void IwyuFileInfo::ReportFullSymbolUse(SourceLocation use_loc,
                                        const char* comment) {
   if (decl) {
     // Since we need the full symbol, we need the decl's definition-site.
-    // But only if we're not defining the function, in which case we want to use
-    // all its preceding declarations, and not try to canonicalize.
-    if (!(flags & UF_FunctionDfn)) {
-      decl = GetDefinitionAsWritten(decl);
+    // Also, by default we canonicalize the location, using GetLocation.
+    const NamedDecl* report_decl = GetDefinitionAsWritten(decl);
+    SourceLocation report_decl_loc = GetLocation(report_decl);
+
+    // However, if we're defining the function or we are targeting an explicit
+    // instantiation, we want to use it as-is and not try to canonicalize at all.
+    if ((flags & (UF_FunctionDfn | UF_ExplicitInstantiation)) != 0) {
```

Comment:
There should be one or more testcases enabled by this change... Any chance you can get them into the same commit?

Reply:
> There should be one or more testcases enabled by this change
You mean, new ones I should write? Without this change the tests wouldn't pass, so it is technically covered.

Reply:
Sorry, I didn't see this before commenting further. It seems to me this change came about for a reason, and it would be nice if that reason was expressed as tests and committed together with the change.

But maybe the tests were failing before this change? Or it makes more sense to commit them with later changes? I think this looks good as-is if it turns out weird with tests together with this.